WebDriverAgentLib/FBApplication.m

```diff
@@ -21,36 +23,70 @@
 @interface FBApplication ()
 @property (nonatomic, assign) BOOL fb_isObservingAppImplCurrentProcess;
 @end
 
 @implementation FBApplication
 
++ (nullable XCAccessibilityElement *)fb_currentAppElement
+{
+  CGPoint screenPoint = CGPointMake(100, 100);
+  __block XCAccessibilityElement *onScreenElement = nil;
+  id<XCTestManager_ManagerInterface> proxy = [FBXCTestDaemonsProxy testRunnerProxy];
+  dispatch_semaphore_t sem = dispatch_semaphore_create(0);
+  [proxy _XCT_requestElementAtPoint:screenPoint
+                              reply:^(XCAccessibilityElement *element, NSError *error) {
+                                if (nil == error) {
+                                  onScreenElement = element;
+                                }
+                                dispatch_semaphore_signal(sem);
+                              }];
+  dispatch_semaphore_wait(sem, dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.3 * NSEC_PER_SEC)));
+  return onScreenElement;
+}
+
 + (instancetype)fb_activeApplication
 {
-  [[[FBRunLoopSpinner new]
-    timeout:5]
+  __block XCAccessibilityElement *currentAppElement = nil;
+  __block XCAccessibilityElement *previousAppElement = self.fb_currentAppElement;
+  if (![[[FBRunLoopSpinner new]
+    timeout:APP_STATE_STABILITY_TIMEOUT]
    spinUntilTrue:^BOOL{
-     return [FBXCAXClientProxy.sharedClient activeApplications].count == 1;
-   }];
+     if ([[[FBRunLoopSpinner new]
+           timeout:APP_STATE_STABILITY_WINDOW]
+          spinUntilTrue:^BOOL{
+            currentAppElement = self.fb_currentAppElement;
+            return !currentAppElement
+              || !previousAppElement
+              || currentAppElement.processIdentifier != previousAppElement.processIdentifier;
+          }]) {
+       previousAppElement = currentAppElement;
+       return NO;
+     }
+     return YES;
+   }]) {
+     [FBLogger logFmt:@"Application state has not been stabilized within %.2f seconds timeout", APP_STATE_STABILITY_TIMEOUT];
+  }
 
   NSArray<XCAccessibilityElement *> *activeApplicationElements = [FBXCAXClientProxy.sharedClient activeApplications];
-  XCAccessibilityElement *activeApplicationElement;
-
-  if (activeApplicationElements.count > 1) {
-    // Might be situations when firstObject is a system application — i.e. SpringBoard
-    XCAccessibilityElement *systemApplicationElement = [FBXCAXClientProxy.sharedClient systemApplication];
-    NSPredicate *nonSystemApplicationPredicate = [NSPredicate predicateWithBlock:^BOOL(id evaluatedObject, NSDictionary<NSString *,id> *bindings) {
-        return systemApplicationElement.processIdentifier != ((XCAccessibilityElement *)evaluatedObject).processIdentifier;
-    }];
-    activeApplicationElement = [[activeApplicationElements filteredArrayUsingPredicate:nonSystemApplicationPredicate] firstObject];
-  } else {
-    activeApplicationElement = [activeApplicationElements firstObject];
+  XCAccessibilityElement *activeApplicationElement = [activeApplicationElements lastObject];
```

Comment:
```activeApplicationElements``` (```activeApplications```) - is unordered, right? why lastObject is better than firstObject ?

Reply:
according to my observations - not really. xctest adds a newly executed app to the tail. Although, the order stays unchanged if one switches between existing apps.